src/Bicep.RoslynAnalyzers/LinterRuleTypeGenerator.cs

```diff
@@ -0,0 +1,196 @@
+            var generatorMethods = initContext.SyntaxProvider
+                .CreateSyntaxProvider(
+                    predicate: (syntax, cancellationToken) => syntax is MethodDeclarationSyntax method && method.AttributeLists.Count > 0,
+                    transform: (syntaxContext, cancellationToken) => TryGetGeneratorMethodContext(syntaxContext, cancellationToken))
+                .Where(method => method is not null);
+
+            var bicepAnalyzerRuleClasses = initContext.SyntaxProvider
+                .CreateSyntaxProvider(
+                    predicate: (syntax, cancellationToken) => syntax is ClassDeclarationSyntax,
+                    transform: (syntaxContext, cancellationToken) => TryGetLinterRuleContext(syntaxContext, cancellationToken))
+                .Where(ruleClass => ruleClass is not null);
+
+            var combined = generatorMethods.Combine(bicepAnalyzerRuleClasses.Collect());
+
+            initContext.RegisterSourceOutput(combined, GenerateSource);
+        }
+
+        
+        private static GeneratorMethodContext? TryGetGeneratorMethodContext(GeneratorSyntaxContext syntaxContext, CancellationToken cancellationToken)
+        {
+            if (syntaxContext.Node is not MethodDeclarationSyntax method)
+            {
+                // shouldn't really happen
+                return null;
+            }
+
+            var attributeSyntax = method.AttributeLists
+                .SelectMany(attributeListSyntax => attributeListSyntax.Attributes)
+                .FirstOrDefault(attributeSyntax => IsGeneratorAttribute(syntaxContext, attributeSyntax, cancellationToken));
+
+            if(attributeSyntax is null)
+            {
+                // did not find our generator attribute
+                return null;
+            }
+
+            var classSyntax = method.FirstAncestorOrSelf<ClassDeclarationSyntax>();
+            if (classSyntax is null)
+            {
+                return null;
+            }
+
+            if (syntaxContext.SemanticModel.GetDeclaredSymbol(classSyntax, cancellationToken: cancellationToken) is not INamedTypeSymbol classSymbol)
+            {
+                return null;
+            }
+
+            return new(method, classSymbol);
+        }
+
+        private static LinterRuleClassContext? TryGetLinterRuleContext(GeneratorSyntaxContext syntaxContext, CancellationToken cancellationToken)
+        {
+            if (syntaxContext.Node is not ClassDeclarationSyntax @class ||
+                syntaxContext.SemanticModel.GetDeclaredSymbol(@class) is not INamedTypeSymbol classSymbol)
+            {
+                // shouldn't happen
+                return null;
+            }
+
+            if(classSymbol.IsAbstract)
+            {
+                // reject abstract classes
+                return null;
+            }
+
+            var ruleTypeSymbol = syntaxContext.SemanticModel.Compilation.GetTypeByMetadataName("Bicep.Core.Analyzers.Interfaces.IBicepAnalyzerRule");
```

Comment:
>GetTypeByMetadataName [](http://example.com/codeflow?start=73&length=21)

We could make the generator more generic by getting the type name from the generator attribute and feeding the `INamedTypeSymbol` through the generator pipeline.

Reply:
The hard coded type makes things fragile because if someone renames `IBicepAnalyzerRule` or changes its namespace, the generator will stop working. I'm going to make this worse by moving the generator to the shared tools repo, so I can't avoid it.

Reply:
Although perhaps we can just put a comment on `IBicepAnalyzerRule` interface not to move or rename it. Most likely, we never will so it won't be a problem.